banyan/src/forest/read.rs

```diff
@@ -19,7 +13,260 @@ use anyhow::{anyhow, Result};
+                            Mode::Forward => {
+                                self.offset += index.count();
+                                self.offset - index.count()..self.offset
+                            }
+                            Mode::Backward => {
+                                self.offset -= index.count();
+                                self.offset..self.offset + index.count()
+                            }
+                        };
+                        head.next_pos(&self.mode);
+                        let placeholder: FilteredChunk<T, V, E> = FilteredChunk {
+                            range,
+                            data: Vec::new(),
+                            extra: (self.mk_extra)(index.as_index_ref()),
+                        };
+
+                        break placeholder;
+                    }
+                }
+
+                Ok(NodeInfo::Leaf(index, leaf)) => {
+                    let chunk = {
+                        if matches!(self.mode, Mode::Backward) {
+                            self.offset -= index.keys.count();
+                        }
+                        let mut matching: SmallVec<[_; 32]> = smallvec![true; index.keys.len()];
+                        self.query.containing(self.offset, index, &mut matching);
+                        let keys = index.select_keys(&matching);
+                        let elems: Vec<V> = match leaf.as_ref().select(&matching) {
+                            Ok(i) => i,
+                            Err(e) => return Some(Err(e)),
+                        };
+                        let offset = self.offset;
+                        let pairs = keys
+                            .zip(elems)
+                            .map(|((o, k), v)| (o + offset, k, v))
+                            .collect::<Vec<_>>();
+                        FilteredChunk {
+                            range: self.offset..self.offset + index.keys.count(),
+                            data: pairs,
+                            extra: (self.mk_extra)(IndexRef::Leaf(index)),
+                        }
+                    };
+                    if matches!(self.mode, Mode::Forward) {
+                        self.offset += index.keys.count();
+                    }
+
+                    // Ascend to parent's node
+                    self.stack.pop();
+                    let last = self.stack.last_mut().expect("not empty");
+                    last.next_pos(&self.mode);
+
+                    break chunk;
+                }
+
+                // even for purged leafs and branches or ignored chunks,
+                // produce a placeholder.
+                //
+                // the caller can find out if we skipped purged parts of the
+                // tree by using an appropriate mk_extra fn, or check
+                // `data.len()`.
+                Ok(_) => {
+                    let TraverseState { index, .. } = self.stack.pop().expect("not empty");
```

Comment:
This assumes that the tree always has at least one branch, correct? With time-based expiry I could imagine a tree that is pruned completely empty.

Reply:
A tree where all events have been pruned will be branch containing a summary of the pruned data, including the offset range. You can never go from a non-empty tree to an empty tree.

The initial empty tree is peeled off with the struct Tree, which is an option of a non empty tree.

Reply:
But - it would be nicer to more explicitly spell out the invariants. So might be worth trying out https://crates.io/crates/contracts for banyan - if there is a way to switch the stuff off without switching to release mode...

Reply:
Made an issue: https://github.com/Actyx/banyan/issues/60

Reply:
In other words: If the root node is a `PrunedBranch`, we will end up in this case with `Ok(PrunedBranch(index))` and everything is fine. This code won't panic.